core/lib/zksync_core/src/state_keeper/cached_storage.rs

```diff
@@ -0,0 +1,270 @@
+        let snapshot_recovery = connection
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await
+            .context("failed getting snapshot recovery info")?;
+        let (miniblock_number, l1_batch_number) = if let Some(snapshot_recovery) = snapshot_recovery
+        {
+            (
+                snapshot_recovery.miniblock_number,
+                snapshot_recovery.l1_batch_number,
+            )
+        } else {
+            (
+                connection
+                    .blocks_dal()
+                    .get_sealed_miniblock_number()
+                    .await?
+                    .unwrap_or_default(),
+                connection
+                    .blocks_dal()
+                    .get_sealed_l1_batch_number()
+                    .await?
+                    .unwrap_or_default(),
+            )
+        };
+
+        tracing::debug!(%l1_batch_number, %miniblock_number, "Using Postgres-based storage");
+        Ok(Box::new(
+            PostgresStorage::new_async(rt_handle, connection, miniblock_number, true).await?,
+        ) as BoxReadStorage)
+    }
+
+    /// Catches up RocksDB synchronously (i.e. assumes the gap is small) and
+    /// returns a [`ReadStorage`] implementation backed by caught-up RocksDB.
+    async fn access_storage_rocksdb<'a>(
+        conn: &mut StorageProcessor<'_>,
+        state_keeper_db_path: &str,
+        enum_index_migration_chunk_size: usize,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<BoxReadStorage<'a>>> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let mut rocksdb_builder = RocksdbStorage::builder(state_keeper_db_path.as_ref())
+            .await
+            .context("Failed initializing RocksDB storage")?;
+        rocksdb_builder.enable_enum_index_migration(enum_index_migration_chunk_size);
+        let rocksdb = rocksdb_builder
+            .synchronize(conn, stop_receiver)
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        if let Some(rocksdb) = &rocksdb {
+            let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+            tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        } else {
+            tracing::warn!("Interrupted")
+        }
+        Ok(rocksdb.map(|rocksdb| Box::new(rocksdb) as BoxReadStorage<'a>))
+    }
+
+    /// Produces a [`ReadStorage`] implementation backed by either Postgres or
+    /// RocksDB (if it's caught up).
+    ///
+    /// # Return value
+    ///
+    /// Returns `Ok(None)` if the process is interrupted using `stop_receiver`.
+    ///
+    /// # Errors
+    ///
+    /// Propagates RocksDB and Postgres errors.
+    pub async fn access_storage(
+        &mut self,
+        rt_handle: Handle,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<BoxReadStorage>> {
```

Comment:
Nit: You should add a lifetime here to make clear that `BoxReadStorage` borrows from `self`. This could make long-term `RwLock` locking more visible as well.